src/NServiceBus.Core/Transports/Learning/DirectoryBasedTransaction.cs

```diff
@@ -0,0 +1,101 @@
+namespace NServiceBus
+{
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Threading.Tasks;
+
+    class DirectoryBasedTransaction : ILearningTransportTransaction
+    {
+        public DirectoryBasedTransaction(string basePath, bool immediateDispatch)
+        {
+            this.basePath = basePath;
+            this.immediateDispatch = immediateDispatch;
+            var transactionId = Guid.NewGuid().ToString();
+
+            transactionDir = Path.Combine(basePath, ".pending", transactionId);
+
+            commitDir = Path.Combine(basePath, ".committed", transactionId);
+        }
+
+        public string FileToProcess { get; private set; }
+
+        public void BeginTransaction(string incomingFilePath)
+        {
+            Directory.CreateDirectory(transactionDir);
+            FileToProcess = Path.Combine(transactionDir, Path.GetFileName(incomingFilePath));
```

Comment:
No strong opinion but it would mean that we would have to start passing that filePath to `InnerProcessFile` and `ProcessFile` so on the fence if it makes the code cleaner?